src/rard/research/views/bibliography.py

```diff
@@ -1,59 +1,231 @@
+
+    def form_valid(self, form):
+        bibliography = form.save(commit=False)
+        updated = form.cleaned_data["antiquarians"]
+        existing = bibliography.antiquarians.all()
+        # get 2 lists of antiquarians, from which to add/remove this bibliography item
+        to_remove = [a for a in existing if a not in updated]
+        to_add = [a for a in updated if a not in existing]
+        for a in to_remove:
+            a.bibliography_items.remove(bibliography)
+        for a in to_add:
+            a.bibliography_items.add(bibliography)
+        c_updated = form.cleaned_data["citing_authors"]
+        c_existing = bibliography.citing_authors.all()
+        # get 2 lists of citing_authors, from which to add/remove this bibliography item
+        to_remove = [c for c in c_existing if c not in updated]
+        to_add = [c for c in c_updated if c not in existing]
+        for c in to_remove:
+            c.bibliography_items.remove(bibliography)
+        for c in to_add:
+            c.bibliography_items.add(bibliography)
+        return super().form_valid(form)
+
+    """def get_context_data(self, *args, **kwargs):
+        context = super().get_context_data(*args, **kwargs)
+        context.update(
+            {
+                "title": self.title,
+                "antiquarians": self.instance.antiquarians.all(),
+            }
+        )
+        return context"""
 
 
 @method_decorator(require_POST, name="dispatch")
 class BibliographyDeleteView(
     CheckLockMixin, LoginRequiredMixin, PermissionRequiredMixin, DeleteView
 ):
-    check_lock_object = "parent"
-
     model = BibliographyItem
-
+    success_url = reverse_lazy("bibliography:list")
     permission_required = ("research.delete_bibliographyitem",)
 
-    def dispatch(self, request, *args, **kwargs):
-        # need to ensure we have the lock object view attribute
-        # initialised in dispatch
-        self.parent = self.get_object().parent
-        return super().dispatch(request, *args, **kwargs)
+    def delete(self, request, *args, **kwargs):
+        """
+        if a TextObjectField contains a mention of a BibilographyItem,
+        it will contain this string:
+        <span class="mention" data-denotation-char="@" data-id="20" data-index="1" data-target="bibliographyitem"  # noqa: E501
+        data-index is the order within the TextObjectField, so we can ignore
+        data-id is the bibliography_item.pk
+        note that the attributes can come in any order, hence the slightly convoluted regex builder (re doesn't support DEFINE)  # noqa: E501
+        """
+        bibliography = self.get_object()
+        bib_pk = bibliography.pk
 
-    def get_success_url(self, *args, **kwargs):
-        return self.parent.get_absolute_url()
+        p = [
+            r"class=\"mention\" ",
+            r"data-denotation-char=\"@\" ",
+            rf"data-id=\"{bib_pk}\" ",
+            # r'data-index="[0-9]{1,4}" ',  # we can ignore this and use .{0,17} instead
```

Comment:
what does `.{0,17}` actually mean in this context?

Reply:
Looks like regular expression syntax for repeat any character 0-17 times.